src/Uno.UI/UI/Xaml/Controls/DatePicker/DatePicker.cs

```diff
@@ -294,46 +361,110 @@ private void BindToFlyout(string propertyName)
+			 * Since we can't just "move" the column, we have to move the textblocks' column, and
+			 * swap the relevant properties (eg: ColumnDefinition.Width) from the old column to the new one. */
+			var orderedColumns = GetOrderedTextBlocksForCulture(CultureInfo.CurrentCulture);
+
+			_flyoutGrid.ColumnDefinitions.Clear();
+
+			_flyoutGrid.ColumnDefinitions.Add(_firstSeparatorColumn);
+			_flyoutGrid.ColumnDefinitions.Add(_secondSeparatorColumn);
+
+			_firstPickerSpacing.Visibility = Visibility.Visible;
+			_secondPickerSpacing.Visibility = Visibility.Visible;
+
+			// Insert columns into the appropriate spaces in between separators
+			int missingColumns = 0;
+			for (var columnIndex = 0; columnIndex < orderedColumns.Length; columnIndex++)
 			{
-				/* DatePicker normally contains 3 textblocks meant for day/month/year with a different Grid.Column each.
-				 * We need to shuffle the columns with the order dictated by current culture, eg: yyyy/mm/dd vs mm/dd/yyyy...
-				 * Since we can't just "move" the column, we have to move the textblocks' column, and
-				 * swap the relevant properties (eg: ColumnDefinition.Width) from the old column to the new one. */
-				var items = GetOrderedTextBlocksForCulture(CultureInfo.CurrentCulture);
-				var oldColumnInfos = new[] { DayColumnPartName, MonthColumnPartName, YearColumnPartName }
-					// TODO: add safe-guard when GetTemplateChild(columnName) is implemented
-					.Select(x => GetColumnIndex(grid, x))
-					.Select(x => new { Column = x, grid.ColumnDefinitions.ElementAtOrDefault(x)?.Width })
-					.ToArray();
-
-				// update TextBlocks' Grid.Column and their respective ColumnDefinition.Width
-				foreach (var item in items)
+				var column = orderedColumns[columnIndex];
+
+				// Check if this column should be displayed based on the TextBlock visibility
+				if (column.Item.Visibility == Visibility.Visible)
+				{
+					var targetPosition = columnIndex * 2 - missingColumns; // Multiply by 2 to adjust for separators
+					_flyoutGrid.ColumnDefinitions.Insert(targetPosition, column.ColumnDefinition);
+					Grid.SetColumn(column.Item, targetPosition);
+				}
+				else
 				{
-					if (grid.ColumnDefinitions.ElementAtOrDefault(oldColumnInfos[item.NewIndex].Column) is ColumnDefinition definition &&
-						oldColumnInfos[item.OldIndex].Width.HasValue)
-					{
-						definition.Width = oldColumnInfos[item.OldIndex].Width.Value;
-					}
-					Grid.SetColumn(item.Item, oldColumnInfos[item.NewIndex].Column);
+					missingColumns++;
 				}
 			}
+
+			var firstSeparatorColumnIndex = _flyoutGrid.ColumnDefinitions.IndexOf(_firstSeparatorColumn);
+			Grid.SetColumn(_firstPickerSpacing, firstSeparatorColumnIndex);
+			var secondSeparatorColumnIndex = _flyoutGrid.ColumnDefinitions.IndexOf(_secondSeparatorColumn);
+			Grid.SetColumn(_secondPickerSpacing, secondSeparatorColumnIndex);
+
+			// Hide first separator if it is the first column of the grid
```

Comment:
can it ever be the last column (or second-last along with the second separator)?

Reply:
How did I miss that case 😅! Indeed, this is possible!